pkg/querier/tripperware/instantquery/instant_query.go

```diff
@@ -0,0 +1,442 @@
+	for _, resp := range resps {
+		if resp.Data.Stats == nil {
+			continue
+		}
+
+		hasStats = true
+		if resp.Data.Stats.Samples == nil {
+			continue
+		}
+
+		for _, s := range resp.Data.Stats.Samples.TotalQueryableSamplesPerStep {
+			if stats, ok := output[s.GetTimestampMs()]; ok {
+				stats.Value += s.Value
+			} else {
+				output[s.GetTimestampMs()] = s
+			}
+		}
+	}
+
+	if !hasStats {
+		return nil
+	}
+
+	return tripperware.StatsMerge(output)
+}
+
+func decorateWithParamName(err error, field string) error {
+	errTmpl := "invalid parameter %q; %v"
+	if status, ok := status.FromError(err); ok {
+		return httpgrpc.Errorf(int(status.Code()), errTmpl, field, status.Message())
+	}
+	return fmt.Errorf(errTmpl, field, err)
+}
+
+// UnmarshalJSON implements json.Unmarshaler.
+func (s *Sample) UnmarshalJSON(data []byte) error {
+	var sample struct {
+		Metric model.Metric    `json:"metric"`
+		Value  cortexpb.Sample `json:"value"`
+	}
+	if err := json.Unmarshal(data, &sample); err != nil {
+		return err
+	}
+	s.Labels = cortexpb.FromMetricsToLabelAdapters(sample.Metric)
+	s.Sample = sample.Value
+	return nil
+}
+
+// MarshalJSON implements json.Marshaler.
+func (s *Sample) MarshalJSON() ([]byte, error) {
+	sample := struct {
+		Metric model.Metric    `json:"metric"`
+		Value  cortexpb.Sample `json:"value"`
+	}{
+		Metric: cortexpb.FromLabelAdaptersToMetric(s.Labels),
+		Value:  s.Sample,
+	}
+	return json.Marshal(sample)
+}
+
+// UnmarshalJSON implements json.Unmarshaler.
+func (s *PrometheusInstantQueryData) UnmarshalJSON(data []byte) error {
+	var queryData struct {
+		ResultType string                               `json:"resultType"`
+		Stats      *tripperware.PrometheusResponseStats `json:"stats,omitempty"`
+	}
+
+	if err := json.Unmarshal(data, &queryData); err != nil {
+		return err
+	}
+	s.ResultType = queryData.ResultType
+	s.Stats = queryData.Stats
+	switch s.ResultType {
+	case model.ValVector.String():
+		var result struct {
+			Samples []*Sample `json:"result"`
+		}
+		if err := json.Unmarshal(data, &result); err != nil {
+			return err
+		}
+		s.Result = PrometheusInstantQueryResult{
+			Result: &PrometheusInstantQueryResult_Vector{Vector: &Vector{
+				Samples: result.Samples,
+			}},
+		}
+	default:
+		s.Result = PrometheusInstantQueryResult{
+			Result: &PrometheusInstantQueryResult_RawBytes{data},
+		}
+	}
+	return nil
+}
+
+// MarshalJSON implements json.Marshaler.
+func (s *PrometheusInstantQueryData) MarshalJSON() ([]byte, error) {
+	switch s.ResultType {
+	case model.ValVector.String():
+		res := struct {
+			ResultType string                               `json:"resultType"`
+			Data       []*Sample                            `json:"result"`
+			Stats      *tripperware.PrometheusResponseStats `json:"stats,omitempty"`
+		}{
+			ResultType: s.ResultType,
+			Data:       s.Result.GetVector().Samples,
+			Stats:      s.Stats,
+		}
+		return json.Marshal(res)
+	default:
+		return s.Result.GetRawBytes(), nil
```

Comment:
This is nice. I should use probably use the same thing on Thanos

Reply:
yeah.... i saw how was done in thanos and this makes it way simpler.